packages/@aws-cdk/toolkit-lib/lib/api/cloud-assembly/private/prepare-source.ts

```diff
@@ -5,122 +5,151 @@ import * as cxschema from '@aws-cdk/cloud-assembly-schema';
- */
-export async function prepareDefaultEnvironment(services: ToolkitServices, props: { outdir?: string } = {}): Promise<Env> {
-  const logFn = (msg: string, ...args: any) => services.ioHelper.notify(IO.CDK_ASSEMBLY_I0010.msg(format(msg, ...args)));
-  const env = await oldPrepare(services.sdkProvider, logFn);
-
-  if (props.outdir) {
-    env[cxapi.OUTDIR_ENV] = props.outdir;
-    await logFn('outdir:', props.outdir);
+export class ExecutionEnviornment {
+  private readonly ioHelper: IoHelper;
+  private readonly sdkProvider: SdkProvider;
+  private readonly debugFn: (msg: string) => Promise<void>;
+  private _outdir: string | undefined;
+
+  public constructor(services: ToolkitServices, props: { outdir?: string } = {}) {
+    this.ioHelper = services.ioHelper;
+    this.sdkProvider = services.sdkProvider;
+    this.debugFn = (msg: string) => this.ioHelper.notify(IO.DEFAULT_ASSEMBLY_DEBUG.msg(msg));
+    this._outdir = props.outdir;
   }
 
-  // CLI version information
-  env[cxapi.CLI_ASM_VERSION_ENV] = cxschema.Manifest.version();
-  env[cxapi.CLI_VERSION_ENV] = versionNumber();
+  /**
+   * Turn the given optional output directory into a fixed output directory
+   */
+  public get outdir(): string {
+    if (!this._outdir) {
+      const outdir = fs.mkdtempSync(path.join(fs.realpathSync(os.tmpdir()), 'cdk.out'));
+      this._outdir = outdir;
+    }
+    return this._outdir;
+  }
 
-  await logFn('env:', env);
-  return env;
-}
+  /**
+   * Guess the executable from the command-line argument
+   *
+   * Only do this if the file is NOT marked as executable. If it is,
+   * we'll defer to the shebang inside the file itself.
+   *
+   * If we're on Windows, we ALWAYS take the handler, since it's hard to
+   * verify if registry associations have or have not been set up for this
+   * file type, so we'll assume the worst and take control.
+   */
+  public guessExecutable(app: string) {
+    return guessExecutable(app, this.debugFn);
+  }
 
-/**
- * Run code from a different working directory
- */
-export async function changeDir<T>(block: () => Promise<T>, workingDir?: string) {
-  const originalWorkingDir = process.cwd();
-  try {
-    if (workingDir) {
-      process.chdir(workingDir);
+  /**
+   * If we don't have region/account defined in context, we fall back to the default SDK behavior
+   * where region is retrieved from ~/.aws/config and account is based on default credentials provider
+   * chain and then STS is queried.
+   *
+   * This is done opportunistically: for example, if we can't access STS for some reason or the region
+   * is not configured, the context value will be 'null' and there could failures down the line. In
+   * some cases, synthesis does not require region/account information at all, so that might be perfectly
+   * fine in certain scenarios.
+   */
+  public async defaultEnvVars(): Promise<Env> {
+    const debugFn = (msg: string) => this.ioHelper.notify(IO.CDK_ASSEMBLY_I0010.msg(msg));
```

Comment:
uses a different code than the other logs. probably doesn't need to. We can make a breaking change at some point.